data/src/main/java/com/eleks/data/firebase/source/impl/FirebaseDataSourceImpl.kt

```diff
@@ -1,9 +1,140 @@
+                        _userGroupsFlow.value = ResultDataModel.error(error.message ?: "")
+                        return@addSnapshotListener
+                    }
+
+                    if (value != null && !value.isEmpty) {
+                        val groups = mutableListOf<GroupDataModel>()
+                        for (doc in value) {
+                            groups.add(doc.toObject())
+                        }
+                        _userGroupsFlow.value = ResultDataModel.success(groups)
+                    }
+                }
+
+        }
+    }
+
+    override fun subscribeSelectedGroups() {
+        if (authInstance.currentUser != null) {
+            dbInstance.collection("group_${authInstance.currentUser?.uid}")
+                .addSnapshotListener { value, error ->
+                    if (error != null) {
+                        _selectedGroupsFlow.value = ResultDataModel.error(error.message ?: "")
+                        return@addSnapshotListener
+                    }
+
+                    if (value != null && !value.isEmpty) {
+                        val selections = mutableListOf<SelectedGroupDataModel>()
+                        for (doc in value) {
+                            selections.add(doc.toObject())
+                        }
+                        _selectedGroupsFlow.value = ResultDataModel.success(selections)
+                    }
+                }
+        }
+    }
+
+    @OptIn(ExperimentalCoroutinesApi::class)
+    override suspend fun saveNewGroup(group: GroupDataModel): ResultDataModel<GroupDataModel> =
+        suspendCancellableCoroutine { continuation ->
+            val uuid = UUID.randomUUID().toString()
+            authInstance.currentUser?.let {
+                val listener = dbInstance.collection("group_${it.uid}").document(uuid)
+                    .addSnapshotListener { _, error ->
+                        if (error != null) {
+                            continuation.resume(ResultDataModel.error(error.message ?: "")) {}
+                            return@addSnapshotListener
+                        }
+                        continuation.resume(ResultDataModel.success(group)) {}
+                    }
+                continuation.invokeOnCancellation { listener.remove() }
+            }
+        }
+
+    override suspend fun saveNewQuote(
+        groupId: String,
+        quote: QuoteDataModel
+    ): ResultDataModel<QuoteDataModel> {
+        TODO("Not yet implemented")
+    }
+
+    @OptIn(ExperimentalCoroutinesApi::class)
+    override suspend fun saveSelection(selectedGroup: SelectedGroupDataModel): ResultDataModel<SelectedGroupDataModel> =
+        suspendCancellableCoroutine { continuation ->
+            val uuid = UUID.randomUUID().toString()
+            authInstance.currentUser?.let {
+                val listener = dbInstance.collection("selection_${it.uid}").document(uuid)
+                    .addSnapshotListener { _, error ->
```

Comment:
I guess adding snapshot listener on every invocation of the 'saveSelection' method. Should we not unsubscribe from it at some point?